core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -632,55 +874,129 @@ public boolean equals(Object o) {
         && Objects.equal(mCachedInode, that.mCachedInode);
   }
 
-  private TraversalResult traverseToInode(String[] pathComponents, boolean collectNonPersisted)
+  private TraversalResult traverseToInode(String[] pathComponents, LockMode lockMode,
+      List<LockMode> lockHints)
       throws InvalidPathException {
     List<Inode<?>> nonPersistedInodes = Lists.newArrayList();
     List<Inode<?>> inodes = Lists.newArrayList();
+    InodeLockGroup lockGroup = new InodeLockGroup();
+
+    // This must be set to true before returning a valid value, otherwise all the inodes will be
+    // unlocked.
+    boolean valid = false;
+    try {
+      if (pathComponents == null) {
+        throw new InvalidPathException("passed-in pathComponents is null");
+      } else if (pathComponents.length == 0) {
+        throw new InvalidPathException("passed-in pathComponents is empty");
+      } else if (pathComponents.length == 1) {
+        if (pathComponents[0].equals("")) {
+          if (lockMode == LockMode.READ) {
+            lockGroup.lockRead(mRoot);
+          } else {
+            lockGroup.lockWrite(mRoot);
+          }
+          inodes.add(mRoot);
+          valid = true;
+          return TraversalResult.createFoundResult(nonPersistedInodes, inodes, lockGroup);
+        } else {
+          throw new InvalidPathException("File name starts with " + pathComponents[0]);
+        }
+      }
 
-    if (pathComponents == null) {
-      throw new InvalidPathException("passed-in pathComponents is null");
-    } else if (pathComponents.length == 0) {
-      throw new InvalidPathException("passed-in pathComponents is empty");
-    } else if (pathComponents.length == 1) {
-      if (pathComponents[0].equals("")) {
-        inodes.add(mRoot);
-        return TraversalResult.createFoundResult(mRoot, nonPersistedInodes, inodes);
+      if (getLockModeForComponent(0, pathComponents.length, lockMode, lockHints) == LockMode.READ) {
```

Comment:
I think this was done to accommodate both traversal from a fresh pathComponents, and also from an existing `inodePath`